amanuensis/resources/admin/admin_associated_user.py

```diff
@@ -25,10 +29,71 @@ def add_associated_users(users):
     with flask.current_app.db.session as session:
         associated_user_schema = AssociatedUserSchema(many=True)
         ret = []
+        users_with_id, users_with_email = [], []
         for user in users:
-            ret.append(udm.add_associated_user(session, user["project_id"], user["email"] if "email" in user else None, user["id"] if "id" in user else None))
-        associated_user_schema.dump(ret)
-        return ret
+            if "email" in user:
+                users_with_email.append(user)
+            elif "id" in user:
+                users_with_id.append(user)
+            else:
+                raise UserError(
+                    "The associated user must have at least an email or id. Neither was found."
+                )
+        user_emails = [user["email"] for user in users_with_email]
+        user_ids = [user["id"] for user in users_with_id]
+        associated_users = udm.get_associated_users(session, user_emails)
+        associated_users += udm.get_associated_users_by_id(session, user_ids)
+
+        fence_users = []
+        if user_emails:
+            fence_users.extend(fence.fence_get_users(config, usernames=user_emails)["users"])
+        if user_ids:
+            fence_users.extend(fence.fence_get_users(config, ids=user_ids)["users"])
+        if fence_users:
+            fence_users = [
+                {"email": user["name"], "id": user["id"]} for user in fence_users
+            ]
 
+        for fence_user in fence_users:
+            for user in users_with_email:
+                if user["email"] == fence_user["email"]:
+                    user["id"] = fence_user["id"]
+                    break
 
+        for fence_user in fence_users:
+            for user in users_with_id:
+                if user["id"] == fence_user["id"]:
+                    user["email"] = fence_user["email"]
+                    break
 
+        associated_user_emails = [user.email for user in associated_users]
+        associated_user_ids = [user.user_id for user in associated_users]
+        
+        seen_id, seen_email = associated_user_ids, associated_user_emails
+
+        users = users_with_email + users_with_id
+        for user in users:
+            email, user_id = None, None
+            if "email" in user:
+                if user["email"] not in seen_email:
+                    seen_email.append(user["email"])
+                    email = user["email"]
+                else:
+                    continue
+            if "id" in user:
+                if user["id"] not in seen_id:
+                    seen_id.append(user["id"])
+                    user_id = user["id"]
+                else:
+                    continue
+
+            ret.append(
+                udm.add_associated_user(
+                    session,
+                    user["project_id"],
+                    email,
+                    user_id,
+                )
+            )
+        associated_user_schema.dump(ret)
```

Comment:
Before this piece of code you :
- Got a user list from the user via API
- Divided in loaded with ID or with email
- Got the fence information about the users (fence_users)

I would refactor this highlighted piece in:
- Divide among present in the associated_user table/variable vs missing from it
    - For the one present, check the values are still current with Fence, if different update
         ```for fence_user in fence_users:
                for user in associated_users:
                    if user["email"] == fence_user["name"] and user["user_id"] != fence_user["id"]:
                          user["user_id"] = fence_user["id"]
                          update_associated_user()```
    - For the missing, create a new associated_user with the combined value from fence and the user input and add it.